### PR TITLE
RqHeaders returns List / Set instead of Iterable

### LINKED ISSUE #490: Refactoring of RqHeaders
Currently RqHeaders has following methods to work with HTTP headers:

    /**
     * Get single header.
     *
     * @param key Header name
     * @return List of values (can be empty)
     * @throws IOException If fails
     */
    Iterable<String> header(CharSequence key) throws IOException;

    /**
     * Get all header names.
     *
     * @return All names
     * @throws IOException If fails
     */
    Iterable<String> names() throws IOException;

However, working with Iterables not really convenient. Example: in order to check if some header exists I have to write something like this:

    new RqHeaders.Base(req)
                        .header(header)
                        .iterator()
                        .hasNext();

Not very clear and obvious. 
At the same time internally RqHeaders implementations contain a kind of multimap - header name to list of values. I think it would be useful to have a couple of additional methods:

    Set<String> namesSet();
    List<String> getHeader(String header);

### PR DESCRIPTION
**Pull request by aliceice**
RqHeaders now has following methods:

```
List<String> header(CharSequence key);
Set<String> names();
```

It makes working with it more convenient.

Fixes #490

### CODE REVIEW

**Comment by davvd** [comment]
@aliceice Thanks, I will ask someone to do a code review here, thanks for your pull request

**Comment by davvd** [comment]
@dmzaytsev it's yours, please review

File: src/main/java/org/takes/rq/RqHeaders.java [hunk 1]
```diff
@@ -90,35 +89,19 @@ public Base(final Request req) {
             super(req);
         }
         @Override
-        public Iterable<String> header(final CharSequence key)
+        public List<String> header(final CharSequence key)
             throws IOException {
             final Map<String, List<String>> map = this.map();
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqHeaders.java]: @aliceice let's get rid once used variable

File: src/main/java/org/takes/rq/RqHeaders.java [hunk 2]
```diff
@@ -90,35 +89,19 @@ public Base(final Request req) {
...
-                    values,
-                    new Sprintf(
-                        "there are only %d headers by name \"%s\"",
-                        values.size(), key
-                    )
-                );
+                return Collections.emptyList();
             }
-            return iter;
+            return values;
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqHeaders.java]: @aliceice please check out this article http://www.yegor256.com/2015/08/18/multiple-return-statements-in-oop.html

**Comment by dmzaytsev** [comment]
@aliceice 
@yegor256 [mention](https://github.com/yegor256/takes/issues/490#issuecomment-169010513)
>  let's use Set and List in header() and names() respectively. 

you use List in header() and Set in names()

**Comment by yegor256** [comment]
@dmzaytsev it's my mistake. should be `List header()` and `Set names()`

**Comment by dmzaytsev** [comment]
@yegor256 got it, thanks

**Comment by dmzaytsev** [comment]
@aliceice @yegor256 another question 
we removed VerboseIterable from header() 
and some diagnostic messages been lost
should header() return something like VerboseLIst now?

**Comment by yegor256** [comment]
@dmzaytsev I think, yes, we need something like `VerboseList`

**Comment by dmzaytsev** [comment]
@aliceice please add a decorator `VerboseList` for `List` like `VerboseIterator`

File: src/main/java/org/takes/rq/RqHeaders.java (hunk 2 above)
Comment (aliceice) [inline src/main/java/org/takes/rq/RqHeaders.java]: Should I reassign `values`instead of returning?

**Comment by aliceice** [comment]
@dmzaytsev Do I understand this right, that I need to implement the `VerboseList`?
I think this should be a separate task. How should I proceed here?

**Comment by dmzaytsev** [comment]
@aliceice yes , you're right
we need to implement `VerboseList` but you can add a [puzzle](http://www.yegor256.com/2009/03/04/pdd.html) for another developer for the implementation `VerboseList` and its tests, for the replacement `List header()` to `VerboseLIst header()` 
Also mention in the puzzle a link to deleted lines 101-116 to save diagnostic messages

**Comment by aliceice** [comment]
@dmzaytsev I updated the PR with the puzzles.

**Comment by aliceice** [comment]
@dmzaytsev Okay I don't know how to structure my puzzle comments to let qulice succeed without checkstyle errors..

File: src/main/java/org/takes/rq/RqHeaders.java [hunk 3]
```diff
@@ -89,36 +88,32 @@
         public Base(final Request req) {
             super(req);
         }
+
+        /**
+         * @todo #490 We need to implement a VerboseList and its tests
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqHeaders.java]: @aliceice `* @todo #490:30min We need to implement a VerboseList and its tests`

File: src/main/java/org/takes/rq/RqHeaders.java [hunk 4]
```diff
@@ -89,36 +88,32 @@
...
-        public Iterable<String> header(final CharSequence key)
+        public List<String> header(final CharSequence key)
             throws IOException {
-            final Map<String, List<String>> map = this.map();
-            final List<String> values = map.get(
+            final List<String> values = this.map().get(
                 key.toString().toLowerCase(Locale.ENGLISH)
             );
-            final Iterable<String> iter;
+            // @todo #490 This list must be wrapped in a VerboseList
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqHeaders.java]: @aliceice please remove this todo block ant put the text into todo above
we don't need two todos on the same ticket

**Comment by aliceice** [comment]
@dmzaytsev Fixed builds :)

**Comment by dmzaytsev** [comment]
@aliceice thank you!

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@elenavolokhova please, check this issue for QA compliance, as per [par.24](http://at.teamed.io/policy.html#24)

**Comment by davvd** [comment]
@rultor deploy pls

**Comment by elenavolokhova** [comment]
@davvd Looks good!

**Comment by davvd** [comment]
> @davvd Looks good!

@elenavolokhova thanks, got it, great to see high-quality tickets here

**Comment by davvd** [comment]
@dmzaytsev added 10 mins to @elenavolokhova, for QA review, payment code is `74515550`... Many thanks! **26 mins** were added to your account in Transaction ID `74515562` (task took 66 hours and 13 mins)... review comments (c=11) added as a bonus... +26 to your rating, your total score is [+1707](http://www.netbout.com/b/35166?open=rating)

**Pull request merged**